### PR TITLE
Added functions to evaluate maxima and minima

### PR DESCRIPTION
**Pull request by SongyuanGuan**
#### References to other Issues or PRs

Implement part of #16520

#### Brief description of what is fixed or changed
In `calculus/util.py`

- Added maxima(): This function returns set of maxima of a function in the given domain.
- Added minima(): This function returns set of minima of a function in the given domain.

In `calculus/tests/test_util.py`
- Added test_maxima(): This function tests the correctness of maxima.
- Added test_minima(): This function tests the correctness of minima.
#### Other comments
1. The maxima and minima currently only work when the function is second order differentiable.
2. In some cases, the function returns correct result but the equal return false. (I commented out line 284 and line 305 in the file sympy/calculus/tests/test_util.) I will be very grateful if anyone could tell me the reason.
```
>>> minima((x+3)*(x-2), x) 
{-1/2}
>>> minima((x+3)*(x-2), x) == {-1/2}
False
```

#### Release Notes

* calculus
    * Added `minima` and `maxima` to find minima and maxima, respectively of a function in a given domain

### CODE REVIEW

**Comment by Sc0rpi0n101** [comment]
There are trailing whitespaces in your edits.
Please remove them first

**Comment by jksuom** [comment]
>  the function is second order differentiable

Should this be checked? SymPy functions are piecewise meromorphic in general. So it would be necessary to consider singularities and discontinuities. Otherwise all derivatives exist.

File: sympy/calculus/util.py [hunk 1]
```diff
@@ -840,6 +840,130 @@ def minimum(f, symbol, domain=S.Reals):
     else:
         raise ValueError("%s is not a valid symbol." % symbol)
 
+def maxima(f, symbol, domain=S.Reals):
+    """
+    Returns the maxima of a function in the given domain.
+
```
Comment (jksuom) [inline sympy/calculus/util.py]: Perhaps it should be explained that these are local maxima, and not even the maximum function values but the argument values where the local maxima are attained.

**Comment by czgdp1807** [comment]
@SongyuanGuan Your work looks quite good. Though, I would like to make one suggestion here which can be here described below,
It would be great if you can reduce the repeated code in `maxima` and `minima` by defining a utility function something like, `_maxmin` and adding options, like `cal_maxima` or `cal_minima`. The interface methods can be still `maxima` and `minima` but they will use `_maxmin` by passing the appropriate options for processing the input. 
I can see that lines, [955] and [958](https://github.com/sympy/sympy/pull/16701/files#diff-a5d2575656fef33ceed5fc4cd005e3c0R958) are the only different ones in both the functions.
My apologies if I misinterpreted your code.

File: sympy/calculus/util.py [hunk 2]
```diff
@@ -840,6 +840,130 @@ def minimum(f, symbol, domain=S.Reals):
...
+    >>> minima(sin(x)*cos(x) + x, x)
+    EmptySet()
+
+    """
+    from sympy import Symbol, diff
+    from sympy.solvers.solveset import solveset
+    from sympy.sets.sets import imageset
+
+    if isinstance(symbol, Symbol):
```
Comment (czgdp1807) [inline sympy/calculus/util.py]: If possible, please put all the checks at the entry point of a function/method.

**Comment by czgdp1807** [comment]
@SongyuanGuan Addressing the point 2 of other comments section.
Try, `minima((x+3)*(x-2), x) == {-S(1)/2}`.

**Comment by oscargus** [comment]
Please consider using the names `argmin` and `argmax` instead as there is no obvious difference between `maximum` and `maxima` etc. I quote Wikipedia:

"In mathematical analysis, the maxima and minima (the respective plurals of maximum and minimum) of a function, known collectively as extrema (the plural of extremum), are the largest and smallest value of the function, either within a given range (the local or relative extrema) or on the entire domain of a function (the global or absolute extrema)."

and

"In mathematics, the arguments of the maxima (abbreviated arg max or argmax) are the points of the domain of some function at which the function values are maximized.[note 1] In contrast to global maxima, referring to the largest outputs of a function, arg max refers to the inputs, or arguments, at which the function outputs are as large as possible. "

**Comment by supreet11agrawal** [comment]
What about discontinuous functions or functions like `abs(x)`?
Take a look at https://github.com/sympy/sympy/blob/6ee4196618e5abcc54f21ec85e78eb5749a341ca/sympy/calculus/util.py#L107
In this function, critical values are defined quite thoroughly. Maybe, this can be used for improvements. And the names should be changed.

**Comment by SongyuanGuan** [comment]
Thanks for all the advice. I am trying to modify the functions to make it work for general conditions.

**Comment by czgdp1807** [comment]
@SongyuanGuan Please refer https://github.com/sympy/sympy/pull/13555 too. It has already done some work on the topic that you are working on.

**Comment by supreet11agrawal** [comment]
I will work on this

**Comment by czgdp1807** [comment]
@SongyuanGuan Are you still working on it?

**Comment by SongyuanGuan** [comment]
I didn’t figure out a good solution for the discontinued condition. And I
am not working on it currently.

**Pull request closed without merging**